internal/terraform/node_action.go

```diff
@@ -0,0 +1,272 @@
+	"github.com/hashicorp/terraform/internal/dag"
+	"github.com/hashicorp/terraform/internal/experiments"
+	"github.com/hashicorp/terraform/internal/lang/langrefs"
+	"github.com/hashicorp/terraform/internal/providers"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+)
+
+// GraphNodeConfigAction is implemented by any nodes that represent an action.
+// The type of operation cannot be assumed, only that this node represents
+// the given resource.
+type GraphNodeConfigAction interface {
+	ActionAddr() addrs.ConfigAction
+}
+
+// nodeExpandActionDeclaration represents an action config block in a configuration module,
+// which has not yet been expanded.
+type nodeExpandActionDeclaration struct {
+	Addr   addrs.ConfigAction
+	Config configs.Action
+
+	Schema           *providers.ActionSchema
+	ResolvedProvider addrs.AbsProviderConfig
+}
+
+var (
+	_ GraphNodeConfigAction      = (*nodeExpandActionDeclaration)(nil)
+	_ GraphNodeReferenceable     = (*nodeExpandActionDeclaration)(nil)
+	_ GraphNodeReferencer        = (*nodeExpandActionDeclaration)(nil)
+	_ GraphNodeDynamicExpandable = (*nodeExpandActionDeclaration)(nil)
+	_ GraphNodeProviderConsumer  = (*nodeExpandActionDeclaration)(nil)
+)
+
+func (n *nodeExpandActionDeclaration) Name() string {
+	return n.Addr.String() + " (expand)"
+}
+
+func (n *nodeExpandActionDeclaration) ActionAddr() addrs.ConfigAction {
+	return n.Addr
+}
+
+func (n *nodeExpandActionDeclaration) ReferenceableAddrs() []addrs.Referenceable {
+	return []addrs.Referenceable{n.Addr.Action}
+}
+
+// GraphNodeModulePath
+func (n *nodeExpandActionDeclaration) ModulePath() addrs.Module {
+	return n.Addr.Module
+}
+
+// GraphNodeAttachActionSchema impl
+func (n *nodeExpandActionDeclaration) AttachActionSchema(schema *providers.ActionSchema) {
+	n.Schema = schema
+}
+
+func (n *nodeExpandActionDeclaration) DotNode(string, *dag.DotOpts) *dag.DotNode {
+	return &dag.DotNode{
+		Name: n.Name(),
+	}
+}
+
+// GraphNodeReferencer
+func (n *nodeExpandActionDeclaration) References() []*addrs.Reference {
+	var result []*addrs.Reference
+	c := n.Config
+
+	refs, _ := langrefs.ReferencesInExpr(addrs.ParseRef, c.Count)
+	result = append(result, refs...)
+	refs, _ = langrefs.ReferencesInExpr(addrs.ParseRef, c.ForEach)
+	result = append(result, refs...)
+
+	if n.Schema != nil && c.Config != nil {
+		refs, _ = langrefs.ReferencesInBlock(addrs.ParseRef, *c.Config, n.Schema.ConfigSchema)
+		result = append(result, refs...)
+	}
+
+	return result
+}
+
+func (n *nodeExpandActionDeclaration) DynamicExpand(ctx EvalContext) (*Graph, tfdiags.Diagnostics) {
+	var g Graph
+	var diags tfdiags.Diagnostics
+	expander := ctx.InstanceExpander()
+	moduleInstances := expander.ExpandModule(n.Addr.Module, false)
+
+	for _, module := range moduleInstances {
+		absActAddr := n.Addr.Absolute(module)
+
+		// Check if the actions language experiment is enabled for this module.
+		moduleCtx := evalContextForModuleInstance(ctx, module)
```

Comment:
FWIW since the parsing of actions is gated behind an experiment, this package won't  "see" any actions associated with the module - there's no need for multiple checks.

Reply:
Couldn't a child module have a self-contained action? I'll add a few test-cases to make sure if this check is needed 👍

Reply:
I think there's two things here. The first one is that the config is just checking if `allowExperiments` is enabled which just means that the build has the `experimentsEnabled=true` [flag](https://github.com/hashicorp/terraform/blob/main/experiments.go#L23) (ie. alpha release). 

There's another thing, which is whether the current module has been opted into a actions specifaclly via the named [actions experiment](https://github.com/hashicorp/terraform/blob/main/internal/experiments/experiment.go#L28).

The difference is basically like, module (or config) authors opt into the second one while the builder of Terraform itself opts into the first one. 

I'm not sure which one we'd want to use - but perhaps the alpha opt in is just enough for now?